CMakeLists.txt

```diff
@@ -0,0 +1,188 @@
+                 examples/environments_on_functions.cpp
+                 examples/functions.cpp
+                 examples/multi_results.cpp
+                 examples/namespacing.cpp
+                 examples/optional_with_iteration.cpp
+                 examples/overloading.cpp
+                 examples/overloading_with_fallback.cpp
+                 examples/protected_functions.cpp
+                 examples/require.cpp
+                 examples/runtime_additions.cpp
+                 examples/script_error_handling.cpp
+                 examples/self_call.cpp
+                 examples/self_from_lua.cpp
+                 examples/stack_aligned_function.cpp
+                 examples/static_variables.cpp
+                 examples/tables.cpp
+                 examples/usertype_advanced.cpp
+                 examples/usertype_automatic_operators.cpp
+                 examples/usertype_bitfields.cpp
+                 examples/usertype_call_from_c++.cpp
+                 examples/usertype.cpp
+                 examples/usertype_dynamic_getter_setter.cpp
+                 examples/usertype_initializers.cpp
+                 examples/usertype_simple.cpp
+                 examples/usertype_special_functions.cpp
+                 examples/usertype_var.cpp
+                 examples/variables.cpp
+                 examples/variadic_args.cpp)
+source_group(EXAMPLES FILES ${EXAMPLES_SRC})
+foreach(example_source_file ${EXAMPLES_SRC})
+message(STATUS "${example_source_file}")
+  get_filename_component(example_name ${example_source_file} NAME_WE)
+  add_executable(${example_name} ${example_source_file} ${HEADER_SRCS})
+  target_link_libraries(${example_name} ${LUA_LIBRARIES})
+  install(TARGETS ${example_name} RUNTIME DESTINATION bin/examples)
+endforeach()
+
+set(TEST_SRC test_state.cpp test_operators.cpp
+             test_tables.cpp test_utility.cpp
+             test_strings.cpp test_environments.cpp
+             test_customizations.cpp test_large_integer.cpp
+             test_inheritance.cpp tests.cpp
+             test_variadics.cpp test_coroutines.cpp
+             test_container_semantics.cpp test_storage.cpp
+             test_overflow.cpp test_plain_types.cpp
+             test_simple_usertypes.cpp test_gc.cpp
+             test_functions.cpp test_usertypes.cpp
+             test_containers.cpp test_filters.cpp)
+source_group(TESTS FILES ${TEST_SRC})
+
+add_executable(tests ${TEST_SRC} ${HEADER_SRCS})
+target_include_directories(tests PRIVATE ./Catch/include/)
+
+find_package(Threads)
+target_link_libraries(tests Threads::Threads ${LUA_LIBRARIES})
+install(TARGETS tests RUNTIME DESTINATION bin)
+
+find_package(PythonInterp 3)
+if (PYTHONINTERP_FOUND)
+  add_custom_command(OUTPUT single/sol.hpp COMMAND ${CMAKE_COMMAND} -E make_directory ${CMAKE_CURRENT_BINARY_DIR}/single && ${PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/single.py --output ${CMAKE_CURRENT_BINARY_DIR}/single/sol.hpp DEPENDS ${HEADER_SRCS})
```

Comment:
shouldn't it be ${CMAKE_CURRENT_SOURCE_DIR}/single/sol.hpp? Some IDE's use a custom build directory, so this will probably break like it is right now.

Reply:
I'm not trying to put it in source directory. I write it in the build directory.
I think that you should never let build tools modify your sources.

Reply:
well, single/sol/sol.hpp is autogenerated, and just for user's conveniance. I don't see the point of having the single sol file automatically generated with cmake if you need to manually copy it over afterwards anyway. Then why don't just execute the command yourself in the first place.

Reply:
hum.
The right way to me seems to make
`
mkdir BUILD && cd BUILD && cmake .. && make install DESTDIR=foo
`
and get the sol.hpp from here.
But it may be convenient to get the single/sol/sol.hpp without compiling it. (windows doesn't have python 3 easily as you said).
So I don't know. But generate source by default in source directory seems me awkward.

Reply:
well, if you compile it with the command line, sure. But consider the fact that you won't compile with the command line on windows, and also on linux alot of people use QtCreator, KDevelop or CLion. The whole point of having it generated in the cmake file is ease of use, having to copy it over afterwards defeats the whole purpose imo.

Reply:
Thinking about it more, maybe you could make it a cmake option. So people who really don't want it auto generated can opt-out of it.